File: src/nfan/fan/sys/RegexMatcher.cs
//
// Copyright (c) 2007, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   27 Dec 07  Andy Frank  Creation
//

using System.Text.RegularExpressions;

namespace Fan.Sys
{
  /// <summary>
  /// RegexMatcher
  /// </summary>
  public sealed class RegexMatcher : FanObj
  {

  //////////////////////////////////////////////////////////////////////////
  // Constructors
  //////////////////////////////////////////////////////////////////////////

    internal RegexMatcher(Match match, string source)
    {
      this.m_match = match;
      this.m_source = source;
    }

  //////////////////////////////////////////////////////////////////////////
  // Identity
  //////////////////////////////////////////////////////////////////////////

    public override Type type() { return Sys.RegexMatcherType; }

  //////////////////////////////////////////////////////////////////////////
  // Methods
  //////////////////////////////////////////////////////////////////////////

    public Boolean matches()
    {
      // to match java
      return (m_match.Success)
        ? Boolean.valueOf(m_source.Length == m_match.Length) : Boolean.False;
    }

    public Boolean find()
    {
      return Boolean.valueOf(true); //matcher.find());
    }

    public long groupCount()
    {
      // to match java
      return m_match.Groups.Count-1;
    }

    public string group() { return group(0); }
    public string group(long group)
    {
      // to match java
      if (!matches().booleanValue()) throw new System.Exception();
      if (group < 0 || group >= m_match.Groups.Count)
        throw IndexErr.make(group).val;

      return m_match.Groups[(int)group].Value;
    }

    public long start() { return start(0); }
    public long start(long group)
    {
      // to match java
      if (!matches().booleanValue()) throw new System.Exception();
      if (group < 0 || group >= m_match.Groups.Count)
        throw IndexErr.make(group).val;

      return m_match.Groups[(int)group].Index;
    }

    public long end() { return end(0); }
    public long end(long group)
    {
      // to match java
      if (!matches().booleanValue()) throw new System.Exception();
      if (group < 0 || group >= m_match.Groups.Count)
        throw IndexErr.make(group).val;

      Group g = m_match.Groups[(int)group];
      return g.Index + g.Length;
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    Match m_match;
    string m_source;

  }
}

File: src/jfan/fan/sys/RegexMatcher.java
//
// Copyright (c) 2007, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   26 Dec 07  Brian Frank  Creation
//
package fan.sys;

import java.util.regex.*;

/**
 * RegexMatcher
 */
public final class RegexMatcher
  extends FanObj
{

//////////////////////////////////////////////////////////////////////////
// Constructors
//////////////////////////////////////////////////////////////////////////

  RegexMatcher(Matcher matcher)
  {
    this.matcher = matcher;
  }

//////////////////////////////////////////////////////////////////////////
// Identity
//////////////////////////////////////////////////////////////////////////

  public Type type() { return Sys.RegexMatcherType; }

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

  public final Boolean matches()
  {
    return matcher.matches();
  }

  public final Boolean find()
  {
    return matcher.find();
  }

  public final Int groupCount()
  {
    return Int.make(matcher.groupCount());
  }

  public final Str group() { return group(Int.Zero); }
  public final Str group(Int group)
  {
    try
    {
      return Str.make(matcher.group((int)group.val));
    }
    catch (IllegalStateException e)
    {
      throw Err.make(e.getMessage()).val;
    }
    catch (IndexOutOfBoundsException e)
    {
      throw IndexErr.make(group).val;
    }
  }

  public final Int start() { return start(Int.Zero); }
  public final Int start(Int group)
  {
    try
    {
      return Int.make(matcher.start((int)group.val));
    }
    catch (IllegalStateException e)
    {
      throw Err.make(e.getMessage()).val;
    }
    catch (IndexOutOfBoundsException e)
    {
      throw IndexErr.make(group).val;
    }
  }

  public final Int end() { return end(Int.Zero); }
  public final Int end(Int group)
  {
    try
    {
      return Int.make(matcher.end((int)group.val));
    }
    catch (IllegalStateException e)
    {
      throw Err.make(e.getMessage()).val;
    }
    catch (IndexOutOfBoundsException e)
    {
      throw IndexErr.make(group).val;
    }
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  Matcher matcher;
}